### DISCUSSION TITLE
Websockets are not working?

### DISCUSSION THREAD

**Discussion by LilTrublMakr** [tobychui/zoraxy#404] [Q&A]
Hello all,

I am having a heck of a time getting websockets to work. All of my services that rely on them (Home Assistant [yes, I have allowed the proxy], NTFY, etc.) are not working. I know in NPM, I needed to explicitly allow websockets but I have read multiple times here that they should automatically be allowed with Zoraxy. Below is the configuration: 

![Screenshot 2024-11-30 090254](https://github.com/user-attachments/assets/f6193c07-c63a-404b-87c6-a306eb260c90)

(Aside: The checkmarks could be darker to be easier to see...)

I am using Technitium for DNS but I don't think that is a problem because if I switch it back to NPM, everything works as it should again.

**Comment by tobychui** [comment]
@LilTrublMakr Is that last slash in your upstream `IP:PORT` intentional? Have you tried removing the trailing slash and see if the websocket works again?

**Comment by LilTrublMakr** [comment]
No, that was not intentional. Removed but still having the same issue. It is across all of my services that need websockets enabled.

**Comment by KernelSailor** [comment]
I have also the Problem that websockets are not working. (For immich and homeassistant)
Has this been solved?
Also for me everythings works just finde if using NPM

**Reply by tobychui** [reply]
Websocket is working on the apps I host in my own homelab. 

For immichi and HA, as I don't self host them, I am not sure if it is compatible. But if it didn't work and you found out why, please feel free to let me know so I can add it to the code / wiki page.

**Reply by KernelSailor** [reply]
Could It be, that the error occurs, because I have two zoraxy instances chained behind each other:

WAN --> zoraxy01 --> zoraxy02 --> WebServer

**Reply by KernelSailor** [reply]
I now finally had the time to test it.
The Problem really is, that there are two reverse Proxys behind each other. When I just use one it works.
But I want to use two behind each other, so that I can have one reverse proxy handle all incomming traffic and then one on each of my docker servers, where I can directly forward the traffic via the docker internal Network.

Do you know what I would need to change in the code or the config, so that websockets work also behind two zoraxy instances?

**Reply by tobychui** [reply]
> But I want to use two behind each other, so that I can have one reverse proxy handle all incomming traffic and then one on each of my docker servers, where I can directly forward the traffic via the docker internal Network.

I am guessing that was caused by Zoraxy is consuming the hop by hop headers required to upgrade to Websocket, so it didn't exactly reproduce your outgoing Websocket request after it passes through the first Zoraxy instance. If your backend require some specific headers to open the Websocket connections (let say there is a special header for auth purpose) , it will not work.

You can try out the development branch (v
3.1.5) with the new upgraded code architecture and Websocket behavior, but generally speaking in cases like this, you need a VPN tunnel instead of two reverse proxy in a role.

**Reply by KernelSailor** [reply]
Thanks @tobychui It works in Version 3.1.5.
Do you know, when this will be meregd into main?

**Reply by tobychui** [reply]
It is already in v3.1.6 main branch. If you want the same behavior as v3.1.5 (which breaks a lot of Proxmox users setup btw), you can turn on websocket header copy in HTTP Proxy > Edit Custom Headers > Advance Settings as shows in the screenshot below

![圖片](https://github.com/user-attachments/assets/30e8f8fd-8ffe-463b-9e34-a3938a31443d)

![圖片](https://github.com/user-attachments/assets/f6416ef9-505e-4266-800b-43c324101f59)
